replica/distributed_loader.cc

```diff
@@ -402,19 +402,31 @@ distributed_loader::make_sstables_available(sstables::sstable_directory& dir, sh
     co_return new_sstables.size();
 }
 
+struct is_shared {
+    using yes = std::true_type;
+    using no = std::false_type;
+};
```

Comment:
How about using seastar::bool_class for this?

Reply:
Botond also had similar suggestion. see https://github.com/scylladb/scylladb/pull/13073#discussion_r1136591150. but for the reason noted in the referenced comment, an instance of `seastar::bool_class` cannot be used as a template parameter.

Reply:
I see.  Well using `sharded` we'll able to get rid of sharing and this won't be required.
That's another indication of the hidden complexity of using std::atomic for sharing across shards.

Reply:
if we agree on using a `sharded<generator>`, can we postpone the change to another PR? because, if the specified `shard_id` is not the one on which the local instance lives, we need to `invoke_on(shard_id)` to get the generation id from its peer. so the `sharded<generator>` change will not only change the generator's `operator()` to return a future, it will also change *all* the caller sites where a synchronized sstable creator is expected. that'd be more much than a refactor. to be more specific, currently, `compaction::_sstable_creator` is the destination of the the assignment chain of the `creator` parameter of `reshard(.., sstables::compaction_sstable_creator_fn creator, ...)` call. and all callers of `compaction::_sstable_creator` returns instance values. to change them and their callers, is non-trivial and risky.

Reply:
Other than the distributed loader and maybe sstable_directory, the other users are already sharded under replica::table so I don't see why they would need to create an sstable for another shard.

Reply:
that said, if we go the sharded way, I agree it deserves a separate PR, but as preparation for this one, not as a follow up.